debian/contrib/manpages/gaminggearfxcontrol.1
.TH GAMINGGEARFXCONTROL "1" "November 2017" "gaminggearfxcontrol version 0.15.1" "User Commands"
.SH NAME
gaminggearfxcontrol \- Control gaminggear fx
.SH DESCRIPTION
.SS "Usage:"
.IP
gaminggearfxcontrol [OPTION?] \- control Gaminggear FX
.SS "Help Options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Show help options
.SS "Application Options:"
.TP
\fB\-d\fR, \fB\-\-device\fR=\fI\,STRING\/\fR
use device STRING
.TP
\fB\-l\fR, \fB\-\-light\fR=\fI\,INDEX\/\fR
use light index INDEX
.TP
\fB\-a\fR, \fB\-\-all\fR
set all colors
.TP
\fB\-s\fR, \fB\-\-single\fR
set single color
.TP
\fB\-r\fR, \fB\-\-reset\fR
reset colors
.TP
\fB\-c\fR, \fB\-\-color\fR
set plain color
.TP
\fB\-1\fR, \fB\-\-color1\fR=\fI\,COLOR1\/\fR
COLOR1


Changelog
0.15.1 2017-06-09

 * Fixed: Translations
 * Fixed: File descriptor handling in GaminggearDevice

0.15.0 2016-09-13

 * Added: GaminggearProfileData, GaminggearProfileTable, GaminggearProfilePage, GaminggearProfiles

0.14.0 2016-08-05

 * Added: gaminggear_hid_to_button_name()
 * Improved: List of HID usage ids

0.13.0 2016-05-22

 * Added: GaminggearHidComboBox
 * Improved: List of HID usage ids
 
0.12.0 2016-04-07

 * Changed: thread related function definitions
 * Fixed two possible segfaults

0.11.2 2015-10-08

 * Fixed compiler warnings
 * Fixed translations
 * Fixed pkg-config problem with missing sqlite3

0.11.1 2015-08-23

 * Fixed compilation problem

0.11.0 2015-08-13

 * Added: Import for Roccat Swarm macro files
 * Fixed: Macroeditor clipboard pasting now handles \t,\n,\r
 * Software implementation of shortcuts/macros now handle multiple events like
   Roccats hardware implementation

0.10.1 2015-06-28

 * Fixed uinput/uhid inclusion problem

0.10.0 2015-06-26

 * Added uhid support for input event handling
 * Fixed install paths for icons and api documentation

0.9.0 2015-05-23

 * Changed uinput handling
 * Improved: library dependencies

0.8.0 2015-04-18

 * Added: Translation de
 * Added: Support for translations
 * Added: GaminggearRWLock
 * Added: definitions for runtime dynamic loading 
 * Improved: gfx supports threading
 * Improved: gaminggearfxcontrol now has more options to control single lights
 * Fixed: pkg-config file

0.7.0 2015-03-10

 * Added: Gfx
 * Fixed: spec file

0.6.0 2015-02-02

 * Improved: detection of cmake-finder install directory
 * Changed: configuration option to build documentation from WITHOUT_DOC to WITH_DOC
 * Changed: libraries now installed in base folder, leaving ld.so search paths umodified
 * Fixed: Missing direct dependencies
 * Fixed: Gtk text != NULL assertions

0.5.0 2014-09-08

 * Improved: GaminggearDevice
 * Fixed: Canberra notifications are failsafe now
 
0.4.0 2014-07-04

 * Added GaminggearAudioNotificator and GaminggearScreenNotificator

0.3.0 2014-06-05

 * Added WITHOUT_DOC build option
 * Added GaminggearDevice
 * Removed libusb dependency

0.2.0 2014-05-18

 * Added GaminggearXyScales
 * Added gaminggearwidget_init()
 * Added TimeoutBar and TimeoutDialog

0.1.0 2014-04-13

 * initial version
 


gaminggearfxcontrol/gaminggearfxcontrol.c
/*
 * This file is part of libgaminggear.
 *
 * libgaminggear is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * libgaminggear is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with libgaminggear. If not, see <http://www.gnu.org/licenses/>.
 */

#include "gaminggear/gfx.h"
#include "config.h"
#include "i18n.h"
#include <glib.h>
#include <locale.h>
#include <stdlib.h>
#include <string.h>

static guint const description_size = 255;

static gchar *parameter_device = NULL;
static gboolean parameter_all = FALSE;
static gboolean parameter_single = FALSE;
static gboolean parameter_reset = FALSE;
static gboolean parameter_color = FALSE;
static gint parameter_light_index = -1;
static gint64 parameter_color1 = 0;

static GOptionEntry entries[] = {
	{ "device", 'd', 0, G_OPTION_ARG_STRING, &parameter_device, N_("use device STRING"), N_("STRING") },
	{ "light", 'l', 0, G_OPTION_ARG_INT, &parameter_light_index, N_("use light index INDEX"), N_("INDEX") },
	{ "all", 'a', 0, G_OPTION_ARG_NONE, &parameter_all, N_("set all colors"), NULL },
	{ "single", 's', 0, G_OPTION_ARG_NONE, &parameter_single, N_("set single color"), NULL },
	{ "reset", 'r', 0, G_OPTION_ARG_NONE, &parameter_reset, N_("reset colors"), NULL },
	{ "color", 'c', 0, G_OPTION_ARG_NONE, &parameter_color, N_("set plain color"), NULL },
	{ "color1", '1', 0, G_OPTION_ARG_INT64, &parameter_color1, N_("COLOR1"), N_("COLOR1") },
	{ NULL }
};

static gboolean post_parse_func(GOptionContext *context, GOptionGroup *group, gpointer data, GError **error) {
	guint actions = 0;

	if (parameter_all) ++actions;
	if (parameter_single) ++actions;
	if (parameter_reset) ++actions;

	if (actions < 1) {
		g_set_error(error, G_OPTION_ERROR, G_OPTION_ERROR_FAILED, _("give one of --all, --single or --reset"));
		return FALSE;
	}

	if (actions > 1) {
		g_set_error(error, G_OPTION_ERROR, G_OPTION_ERROR_FAILED, _("--all, --single and --reset are mutual exclusive"));
		return FALSE;
	}

	if ((parameter_all || parameter_single) && !parameter_color) {
		g_set_error(error, G_OPTION_ERROR, G_OPTION_ERROR_FAILED, _("--all and --single need --color"));
		return FALSE;
	}

	if (parameter_single && (!parameter_device || parameter_light_index == -1)) {
		g_set_error(error, G_OPTION_ERROR, G_OPTION_ERROR_FAILED, _("--single needs --device and --light"));
		return FALSE;
	}

	if (parameter_color && !parameter_color1) {
		g_set_error(error, G_OPTION_ERROR, G_OPTION_ERROR_FAILED, _("--color needs --color1"));
		return FALSE;
	}

	return TRUE;
}

static GOptionContext *commandline_parse(int *argc, char ***argv) {
	GOptionContext *context = NULL;
	GError *error = NULL;

	context = g_option_context_new(_("- control Gaminggear FX"));

	g_option_context_add_main_entries(context, entries, NULL);
	g_option_context_set_translation_domain(context, PROJECT_NAME);

	g_option_group_set_parse_hooks(g_option_context_get_main_group(context), NULL, post_parse_func);

	if (!g_option_context_parse(context, argc, argv, &error)) {
		g_critical(_("Option parsing failed: %s"), error->message);
		exit(EXIT_FAILURE);
	}
	return context;
}

static void commandline_free(GOptionContext *context) {
	g_free(parameter_device);
}

static gint get_device_index(gchar const *parameter) {
	GfxResult gfx_result;
	unsigned int num_devices;
	unsigned int device_index;
	char description[description_size];
	GfxDevtype device_type;

	gfx_result = gfx_get_num_devices(&num_devices);
	if (gfx_result != GFX_SUCCESS) {
		g_warning(_("Could not get devices"));
		return -1;
	}

	for (device_index = 0; device_index < num_devices; ++device_index) {
		gfx_result = gfx_get_device_description(device_index, description, description_size, &device_type);
		if (gfx_result != GFX_SUCCESS) {
			g_warning(_("Could not get device description"));
			return -1;
		}

		if (!strcmp(description, parameter))
			return device_index;
	}
	return -1;
}

int main(int argc, char **argv) {
	GOptionContext *context;
	GfxResult gfx_result;
	int retval = EXIT_FAILURE;
	gint device_index;

	setlocale(LC_ALL, "");
	bindtextdomain(PROJECT_NAME, LOCALEDIR);
	bind_textdomain_codeset(PROJECT_NAME, "UTF-8");
	textdomain(PROJECT_NAME);

	context = commandline_parse(&argc, &argv);

	gfx_result = gfx_initialize();
	if (gfx_result != GFX_SUCCESS) {
		g_warning(_("Could not initialize fx system"));
		goto exit_1;
	}

	if (parameter_single) {
		device_index = get_device_index(parameter_device);
		if (device_index == -1) {
			g_warning(_("Could not find device %s"), parameter_device);
			goto exit_2;
		}
		if (parameter_color) {
			gfx_result = gfx_set_light_color(device_index, parameter_light_index, parameter_color1);
			if (gfx_result != GFX_SUCCESS) {
				g_warning(_("Could not set light %1$i on device %2$s to color1"), parameter_light_index, parameter_device);
				goto exit_2;
			}
		}
	} else if (parameter_all) {
		if (parameter_color) {
			gfx_result = gfx_light(GFX_LOCATION_ALL, parameter_color1);
			if (gfx_result != GFX_SUCCESS) {
				g_warning(_("Could not set all lights to color1"));
				goto exit_2;
			}
		}
	} else if (parameter_reset) {
		gfx_result = gfx_reset();
		if (gfx_result != GFX_SUCCESS) {
			g_warning(_("Could not reset all lights"));
			goto exit_2;
		}
	}

	if (parameter_all || parameter_single || parameter_reset) {
		gfx_result = gfx_update();
		if (gfx_result != GFX_SUCCESS) {
			g_warning(_("Could not update lights"));
			goto exit_2;
		}
	}

	retval = EXIT_SUCCESS;
exit_2:
	gfx_result = gfx_release();
	if (gfx_result != GFX_SUCCESS) {
		g_warning(_("Could not release fx system"));
		retval = EXIT_FAILURE;
		goto exit_1;
	}
exit_1:
	commandline_free(context);
	return retval;
}
